src/Nethermind/Nethermind.Db/SimpleFilePublicKeyDb.cs

```diff
@@ -175,24 +179,97 @@ public void Dispose()
             }
 
-            string[] lines = File.ReadAllLines(DbPath);
-            foreach (string line in lines)
+            using SafeFileHandle fileHandle = File.OpenHandle(DbPath, FileMode.OpenOrCreate);
+
+            byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(maxLineLength);
+            int read = RandomAccess.Read(fileHandle, rentedBuffer, 0);
+
+            long offset = 0L;
+            Span<byte> bytes = default;
+            while (read > 0)
             {
-                string[] values = line.Split(",");
-                if (values.Length != 2)
+                offset += read;
+                bytes = rentedBuffer.AsSpan(0, read + bytes.Length);
+                while (true)
                 {
-                    if (_logger.IsError) _logger.Error($"Error when loading data from {Name} - expected two items separated by a comma and got '{line}')");
-                    continue;
+                    // Store the original span incase need to undo the key slicing if end of line not found
+                    Span<byte> iterationSpan = bytes;
+                    int commaIndex = bytes.IndexOf((byte)',');
+                    Span<byte> key = default;
+                    if (commaIndex >= 0)
+                    {
+                        key = bytes[..commaIndex];
+                        bytes = bytes[(commaIndex + 1)..];
+                    }
+                    int lineEndIndex = bytes.IndexOf((byte)'\n');
+                    if (lineEndIndex < 0)
+                    {
+                        // Restore the iteration start span
+                        bytes = iterationSpan;
+                        break;
+                    }
+
+                    Span<byte> value;
+                    if (bytes[lineEndIndex - 1] == (byte)'\r')
+                    {
+                        // Windows \r\n
+                        value = bytes[..(lineEndIndex - 1)];
+                    }
+                    else
+                    {
+                        // Linux \n
+                        value = bytes[..lineEndIndex];
+                    }
+
+                    if (commaIndex < 0)
+                    {
+                        // End of line but no comma
+                        RecordError(value);
+                    }
+                    else if (lineEndIndex >= 0)
+                    {
+                        _cache[Bytes.FromUtf8HexString(key)] = Bytes.FromUtf8HexString(value);
+                    }
+                    // Move to after end of line
+                    bytes = bytes[(lineEndIndex + 1)..];
+                }
+
+                if (bytes.Length > 0)
+                {
+                    // Move up any remaining to start of buffer
+                    bytes.CopyTo(rentedBuffer);
+                }
+
+                read = RandomAccess.Read(fileHandle, rentedBuffer.AsSpan(bytes.Length), offset);
+            }
+
+            ArrayPool<byte>.Shared.Return(rentedBuffer);
```

Comment:
should it be in `finally` clause?

Reply:
`ArrayPool.Shared` is more like a cache; so will happily create a new array and old one will be GC'd; and the cost of an exception being thrown will dwarf the cost of a new 2kB array being created quite significantly, so it doesn't matter too much if its returned.

Crossing try/catch/finally boundaries does come at a small cost of forcing any variables to the stack rather than registers (so they can be restored after the exception is caught)

More to the point though; if an exception is thrown here the process will exit (as I found during testing) so it doesn't really matter that the array is returned either 😅